src/DocumentFormat.OpenXml/Linq/LinqExtensions.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using DocumentFormat.OpenXml.Packaging;
+using System;
+using System.IO;
+using System.Xml;
+using System.Xml.Linq;
+
+namespace DocumentFormat.OpenXml.Linq
+{
+    /// <summary>
+    /// Provides extension methods for reading and writing <see cref="XDocument" />
+    /// and <see cref="XElement" /> instances from and to <see cref="OpenXmlPart"/>
+    /// instances.
+    /// </summary>
+    public static class LinqExtensions
+    {
+        internal static readonly XDeclaration XDeclaration = new("1.0", "UTF-8", "yes");
+
+        /// <summary>
+        /// Gets the given <see cref="OpenXmlPart" />'s root <see cref="XDocument" />.
+        /// </summary>
+        /// <param name="part">The <see cref="OpenXmlPart" />.</param>
+        /// <returns>The root <see cref="XDocument" />.</returns>
+        public static XDocument GetXDocument(this OpenXmlPart part)
+        {
+            var rootDocument = part.Annotation<XDocument>();
+            if (rootDocument != null)
+            {
+                return rootDocument;
+            }
+
+            using Stream stream = part.GetStream();
+            if (stream.Length > 0)
+            {
+                using XmlReader xmlReader = XmlReader.Create(stream);
+                rootDocument = XDocument.Load(xmlReader);
+            }
+            else
+            {
+                if (LinqUtility.OpenXmlPartInfos.TryGetValue(part.GetType(), out OpenXmlPartInfo? partInfo))
```

Comment:
Check out the `part.Metadata` API - it should have much of what you're looking for (can't remember the exact name, though)

Reply:
Do you mean the classes in the `Framework\Metadata` folder? There is a class called `ElementLookup` which seems to provide that information. It seems to require a deeper look, though.